### DISCUSSION TITLE
How to create sdist build with specified providers?

### DISCUSSION THREAD

**Discussion by aksswami** [apache/airflow#15329] [Q&A] [answered]
I have airflow fork from 1.10.14 and I was building sdist previously using

`python setup.py compile_assets sdist --formats=gztar`

But in airflow 2.0, the way provider packages are packaged, the above command will not work. Also not sure how to configure which packages to include using the above.

If I set`INSTALL_PROVIDERS_FROM_SOURCES=true` to include provider source, but get_provider_info.py not created and linking of provider is also not done. So when running airflow getting an issue that `apache_airflow_provider_*` not found.

Any help would be great. I have tried to check scripts, which build airflow. But most are using breeze, which again builds docker image. I am not using docker as of now.

Thanks in advance.

**Comment by potiuk** [comment, marked as answer]
The `python setup.py compile_assets sdist --formats=gztar`  will work, but It will produce only airflow package (no providers).

If you want to add providers, you can build them separately using breeze.

The instructions on how to do it are available here: https://github.com/apache/airflow/blob/master/dev/PROVIDER_PACKAGE_DETAILS.md#preparing-packages

Then you need to install both `airflow` and selected `providers` package. They use breeze and this is intended, because while building providers we are veifying if the providers are properly importable and it uses certain dependencies of python to generate the setup.py dynmically.

You can also use sdist providers of the provider packages (and airflow) from the ones released by the community: 

Providers: https://downloads.apache.org/airflow/providers/
Airflow: https://downloads.apache.org/airflow/2.0.1/

This is the best way to proceed as you can verify signatures/checksums for those - those are the only "official" releases of airflow.

However if you **REALLY** want to build the provider packages on your own you can follow:

https://github.com/apache/airflow/blob/master/dev/provider_packages/README.md#preparing-provider-packages

The chapter describes how to use breeze in docker (as it provides you the common dev environment) however if you are adventurous the next chapter: 

https://github.com/apache/airflow/blob/master/dev/provider_packages/README.md#testing-and-debugging-provider-preparation 

Explains step-by-step how to get to generating your own package using your local virtualenv, without docker (I use it for testing and debugging). You can follow the commands specified there.

**Comment by potiuk** [comment]
Maybe - while doing it -  you can extract it out to a common set of instructions that should be run to get your own providers. I did not realise this is really needed, but maybe you could prepare such set of instructions and share it with the community. Such instructions have the biggest value when contributed from a user, as a lot of assumptions that I have in my head (and other contributors) are not there - and it's a great way to give back to the community!

**Reply by aksswami** [reply]
Thanks @potiuk for the answer. I have used the prebuild sdist from the community for the providers as of now to unblock. I will check the `breeze` and  `python ./dev/provider_packages/prepare_provider_packages.py --packages` command in details later.

I will surely update here or share steps once get this working.

Can we keep this thread open, in case need to discuss some command option `prepare_provider_packages` provides.?

**Reply by potiuk** [reply]
Sure
